src/network/bufferedconnection.h

```diff
@@ -0,0 +1,321 @@
+		 */
+		bool uint8_t(uint8_t* out = nullptr);
+
+		/**
+		 * Checks whether a std::string can be read from the buffer.
+		 * This method does not modify the buffer contents but increases the peek-pointer.
+		 * \return \c True if the value can be read, \c false if not enough data has been received.
+		 */
+		bool string();
+
+		/**
+		 * Checks whether a RecvPacket can be read from the buffer.
+		 * This method does not modify the buffer contents but increases the peek-pointer.
+		 * \return \c True if the value can be read, \c false if not enough data has been received.
+		 */
+		bool recvpacket();
+
+	private:
+		/// The connection to operate on.
+		BufferedConnection* conn_;
+
+		/// The position of the next peek.
+		size_t peek_pointer_;
+	};
+
+	/**
+	 * Tries to establish a connection to the given address.
+	 * \param host The host to connect to.
+	 * \return A pointer to a connected \c BufferedConnection object or a \c nullptr.
+	 */
+	static std::unique_ptr<BufferedConnection> connect(const NetAddress& host);
+
+	/**
+	 * Tries to create a connection offered by the given acceptor.
+	 * \param host An acceptor which might have a pending connection.
+	 * \return A pointer to a connected \c BufferedConnection object or a \c nullptr.
+	 */
+	static std::unique_ptr<BufferedConnection> accept(boost::asio::ip::tcp::acceptor& acceptor);
+
+	/**
+	 * Closes the connection.
+	 * If you want to send a goodbye-message, do so before freeing the object.
+	 */
+	~BufferedConnection();
+
+	/**
+	 * Returns whether the connection is established.
+	 * \return \c true if the connection is open, \c false otherwise.
+	 */
+	bool is_connected() const;
+
+	/**
+	 * Closes the connection.
+	 * If you want to send a goodbye-message, do so before calling this.
+	 */
+	void close();
+
+	/**
+	 * Receive a command.
+	 * \warning Calling this method is only safe when peek_cmd() returned \c true.
+	 *          Otherwise the behavior of this method is undefined.
+	 * \param out The variable to write the value to.
+	 */
+	void receive(RelayCommand* out);
+
+	/**
+	 * Receive an uint8_t.
+	 * \warning Calling this method is only safe when peek_uint8_t() returned \c true.
+	 *          Otherwise the behavior of this method is undefined.
+	 * \param out The variable to write the value to.
+	 */
+	void receive(uint8_t* out);
+
+	/**
+	 * Receive a string.
+	 * \warning Calling this method is only safe when peek_string() returned \c true.
+	 *          Otherwise the behavior of this method is undefined.
+	 * \param out The variable to write the value to.
+	 */
+	void receive(std::string* out);
+
+	/**
+	 * Receive a RecvPacket.
+	 * \warning Calling this method is only safe when peek_recvpacket() returned \c true.
+	 *          Otherwise the behavior of this method is undefined.
+	 * \param out The variable to write the value to.
+	 */
+	void receive(RecvPacket* out);
+
+	// Temporary method, will be removed.
```

Comment:
will be removed when / after ?

Reply:
When the data is used somewhere. I extended the comment.